### ISSUE TITLE
custom_target places output in wrong folder

### ISSUE THREAD

**Issue by kaimast** [mesonbuild/meson#1765]
I have a nested meson setup: In ./ i have a meson.build with "subdir(./src/server/meson.build')"

The meson.build in src/server contains a custom target. However the output of the custom target is placed in ./build instead of ./build/src/server.

Is this a known issue? Any idea on how to work around it? I tested on both meson in ubuntu-zesty and meson built myself from this github.

**Comment by jpakkane** [comment]
Please post a link to your full project, this really can't be debugged based on text description alone.

**Comment by kaimast** [comment]
Try this one https://github.com/kaimast/meson-bug

**Comment by nirbheek** [comment]
Meson can't magically make your tools place outputs in the right folders. :)

The way you would use it is:

in the root dir:

```meson
gen_file = files('gen-file.sh')
```

```meson
gen_source = custom_target('gen-source',
                           output: ['myfunc.c'],
                           command : [gen_file, '@OUTPUT@'])
```

and `gen-file.sh` should read `$1` as the file to write to. The `files()` part is optional, of course. It's just an easy way to tell meson what file you refer to and let it handle translating paths regardless of what subdir you are in.

**Comment by kaimast** [comment]
I don't think you understand the problem. The tool places the output in the right folder ./src/server/, but then meson copies it to ./build/ instead of ./build/src/server.

This is just a toy example. I cannot modify the tool to handle paths. I just uses the current working directroy. The tool is also not part of the repo in my real setup (that I cannot share here unfortunately).

The output looks like that:
`kai@chipad:~/Development/meson-bug/build$ ninja
[2/3] Compiling C object 'src/server/testfile@exe/src_server_myfunc.c.o'.
FAILED: src/server/testfile@exe/src_server_myfunc.c.o 
cc  '-Isrc/server/testfile@exe' '-Isrc/server/' '-I../src/server' '-fdiagnostics-color=always' '-pipe' '-D_FILE_OFFSET_BITS=64' '-Wall' '-Winvalid-pch' '-O0' '-g' '-MMD' '-MQ' 'src/server/testfile@exe/src_server_myfunc.c.o' '-MF' 'src/server/testfile@exe/src_server_myfunc.c.o.d' -o 'src/server/testfile@exe/src_server_myfunc.c.o' -c src/server/myfunc.c
cc: error: src/server/myfunc.c: No such file or directory
cc: fatal error: no input files
compilation terminated.
ninja: build stopped: subcommand failed.`

Related question, why is there no option to set the working directory for custom_target?

**Comment by kaimast** [comment]
Realized I should have been using generators for my usecase, but it still doesn't seem to work.

(Sorry I accidentally commented on the wrong issue...)

**Comment by jpakkane** [comment]
Setting the cwd is not currently supported because it is difficult to guarantee on all backends and would also require wrapping _every_ command invocation with a wrapper script which is slow.

This is unfortunate but tools that can't be told where to write their output are bad, and integrating them into any build system (except the exact one that the original developer had) is going to be painful no matter what. The fastest way to solve this is to write a wrapper script that takes arguments and does the `cd` magic and other things necessary to set up the environment for the actual script.

**Comment by kaimast** [comment]
Just to make clear, the problem is that it meson enforces out of tree buildirectories. 

So you're saying I should cd into the build directory and then run the script?

**Comment by kaimast** [comment]
Okay that actually works... I still think the working directory should be the build directory by default. But I guess you can close this.

Thanks for the help :)

**Comment by nioncode** [comment]
@jpakkane Please close.
